Writing to a DBR_LONG PV from a Text Entry widget results in an exception when using the EPICS Jackie backend

In Phoebus 5.0.2, there is a problem when using the EPICS Jackie backend and writing to a PV of type `DBR_LONG` from a Text Entry widget (and potentially other widgets). The write operation fails with an exception:

```
025-10-23 15:31:58 WARNING [org.csstudio.display.builder.runtime] Widget Text Entry write error for value 2
java.lang.Exception: Cannot write 2 to PV TEST:Longout
	at org.csstudio.display.builder.runtime.pv.RuntimePV.write(RuntimePV.java:114)
	at org.csstudio.display.builder.runtime.WidgetRuntime.writePrimaryPV(WidgetRuntime.java:381)
	at org.csstudio.display.builder.runtime.RuntimeUtil$1.handleWrite(RuntimeUtil.java:52)
	at org.csstudio.display.builder.representation.ToolkitRepresentation.fireWrite(ToolkitRepresentation.java:682)
	at org.csstudio.display.builder.representation.javafx.widgets.TextEntryRepresentation.submit(TextEntryRepresentation.java:247)
	at org.csstudio.display.builder.representation.javafx.widgets.TextEntryRepresentation.lambda$createJFXNode$0(TextEntryRepresentation.java:141)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:232)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:189)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene.processKeyEvent(Scene.java:2194)
	at javafx.scene.Scene$ScenePeerListener.keyEvent(Scene.java:2715)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:218)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:150)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$1(GlassViewEventHandler.java:250)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:424)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(GlassViewEventHandler.java:249)
	at com.sun.glass.ui.View.handleKeyEvent(View.java:542)
	at com.sun.glass.ui.View.notifyKey(View.java:966)
Caused by: java.lang.IllegalArgumentException: Cannot convert object of type java.lang.Long: 2
	at org.phoebus.pv.jackie.util.ValueConverter.objectToChannelAccessSimpleOnlyValue(ValueConverter.java:399)
	at org.phoebus.pv.jackie.JackiePV.write(JackiePV.java:306)
	at org.csstudio.display.builder.runtime.pv.RuntimePV.write(RuntimePV.java:110)
	... 42 more
```

The problem is that `FormatOptionHandler` generates an object of type `Long`, but the `ValueConverter` of the EPICS Jackie PV provider currently only handled `Byte`, `Short`, and `Integer`. While the Channel Access protocol does not support 64-bit integer values (as represented by the Java `Long` type), the PV provider should be able to handle the `Long` type as long as the actual value fits into the limits of the `Integer` type. It could even be argued that values outside this limit should be supported, implicitly converting them to `DBR_DOUBLE` and thus losing precision, like the JCA provider does.

I am going to take care of the necessary changes to the code.

--- a/core/pv-jackie/pom.xml
+++ b/core/pv-jackie/pom.xml
@@ -23,6 +23,13 @@
     </dependency>
 
     <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <version>${mockito.version}</version>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
       <groupId>org.epics</groupId>
       <artifactId>vtype</artifactId>
       <version>${vtype.version}</version>

--- a/core/pv-jackie/src/main/java/org/phoebus/pv/jackie/JackiePV.java
+++ b/core/pv-jackie/src/main/java/org/phoebus/pv/jackie/JackiePV.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2017-2024 aquenos GmbH.
+ * Copyright (c) 2017-2025 aquenos GmbH.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -261,10 +261,12 @@ public class JackiePV extends PV {
                 // Use ca_put_callback.
                 final var listenable_future = channel.put(
                         ValueConverter.objectToChannelAccessSimpleOnlyValue(
+                                channel.getName(),
                                 new_value,
                                 channel.getClient().getConfiguration()
                                         .getCharset(),
-                                treat_char_as_long_string));
+                                treat_char_as_long_string,
+                                preferences.long_conversion_mode()));
                 final var completable_future = new CompletableFuture<Void>();
                 listenable_future.addCompletionListener((future) -> {
                     try {
@@ -304,10 +306,12 @@ public class JackiePV extends PV {
                 // Use ca_put without a callback.
                 channel.putNoCallback(
                         ValueConverter.objectToChannelAccessSimpleOnlyValue(
+                                channel.getName(),
                                 new_value,
                                 channel.getClient().getConfiguration()
                                         .getCharset(),
-                                treat_char_as_long_string));
+                                treat_char_as_long_string,
+                                preferences.long_conversion_mode()));
             }
             case YES -> {
                 // Wait for the write operation to complete. This allows code

--- a/core/pv-jackie/src/main/java/org/phoebus/pv/jackie/JackiePreferences.java
+++ b/core/pv-jackie/src/main/java/org/phoebus/pv/jackie/JackiePreferences.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2024 aquenos GmbH.
+ * Copyright (c) 2024-2025 aquenos GmbH.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -133,9 +133,72 @@ public record JackiePreferences(
         boolean dbe_property_supported,
         boolean honor_zero_precision,
         String hostname,
+        LongConversionMode long_conversion_mode,
         ChannelAccessEventMask monitor_mask,
         boolean rtyp_value_only,
         String username) {
+
+    /**
+     * Mode for handling integer numbers that are too large to fit into the
+     * {@link Integer} type.
+     */
+    public enum LongConversionMode {
+        /**
+         * Limit the {@link Long} value to the {@link Integer} range.
+         *
+         * A value that is greater than {@link Integer#MAX_VALUE} is converted
+         * to {@link Integer#MAX_VALUE}, and a value that is less than
+         * {@link Integer#MIN_VALUE} is converted to {@link Integer#MIN_VALUE}.
+         */
+        COERCE,
+
+        /**
+         * Limit the {@link Long} value to the {@link Integer} range and log a
+         * warning.
+         *
+         * This essentially is the same behavior as {@link #COERCE}, but a
+         * warning message is written to the log.
+         */
+        COERCE_AND_WARN,
+
+        /**
+         * Convert the {@link Long} value to a {@link Double}.
+         *
+         * This means that some precision is lost.
+         */
+        CONVERT,
+
+        /**
+         * Convert the {@link Long} value to a {@link Double} and log a
+         * warning.
+         *
+         * This essentially is the same behavior as {@link #CONVERT}, but a
+         * warning message is written to the log.
+         */
+        CONVERT_AND_WARN,
+
+        /**
+         * Raise an {@link IllegalArgumentException}.
+         */
+        FAIL,
+
+        /**
+         * Cast the {@link Long} value to an {@link Integer}.
+         *
+         * This means that the value will overflow (e.g.
+         * <code>Integer.MAX_VALUE + 1</code> becomes
+         * <code>Integer.MIN_VALUE</code>.
+         */
+        TRUNCATE,
+
+        /**
+         * Cast the {@link Long} value to an {@link Integer} and log a warning.
+         *
+         * This essentially is the same behavior as {@link #CAST}, but a
+         * warning message is written to the log.
+         */
+        TRUNCATE_AND_WARN,
+    }
 
     private final static JackiePreferences DEFAULT_INSTANCE;
 
@@ -215,6 +278,18 @@ public record JackiePreferences(
         var hostname = preference_reader.get("hostname");
         if (hostname.isEmpty()) {
             hostname = null;
+        }
+        final var long_conversion_mode_string = preference_reader.get(
+                "long_conversion_mode");
+        LongConversionMode long_conversion_mode;
+        try {
+            long_conversion_mode = LongConversionMode.valueOf(
+                    long_conversion_mode_string);
+        } catch (IllegalArgumentException|NullPointerException e) {
+            logger.severe(
+                    "Invalid long conversion mode: "
+                            + long_conversion_mode_string);
+            long_conversion_mode = LongConversionMode.COERCE_AND_WARN;
         }
         final var monitor_mask_string = preference_reader.get("monitor_mask");
         ChannelAccessEventMask monitor_mask;
@@ -493,6 +568,7 @@ public record JackiePreferences(
         logger.config("dbe_property_supported = " + dbe_property_supported);
         logger.config("honor_zero_precision = " + honor_zero_precision);
         logger.config("hostname = " + hostname);
+        logger.config("long_conversion_mode = " + long_conversion_mode);
         logger.config("monitor_mask = " + monitor_mask);
         logger.config("rtyp_value_only = " + rtyp_value_only);
         logger.config("use_env = " + use_env);
@@ -513,6 +589,7 @@ public record JackiePreferences(
                 dbe_property_supported,
                 honor_zero_precision,
                 hostname,
+                long_conversion_mode,
                 monitor_mask,
                 rtyp_value_only,
                 username);

--- a/core/pv-jackie/src/main/java/org/phoebus/pv/jackie/util/ValueConverter.java
+++ b/core/pv-jackie/src/main/java/org/phoebus/pv/jackie/util/ValueConverter.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2024 aquenos GmbH.
+ * Copyright (c) 2024-2025 aquenos GmbH.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -52,6 +52,9 @@ import org.epics.vtype.VString;
 import org.epics.vtype.VStringArray;
 import org.epics.vtype.VType;
 import org.phoebus.core.vtypes.VTypeHelper;
+import org.phoebus.pv.jackie.JackiePreferences;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.nio.ByteBuffer;
 import java.nio.DoubleBuffer;
@@ -74,6 +77,14 @@ public final class ValueConverter {
      * Offset between the UNIX and EPICS epoch in seconds.
      */
     public static final long OFFSET_EPICS_TO_UNIX_EPOCH_SECONDS = 631152000L;
+
+    /**
+     * Logger used by this class.
+     *
+     * The field is package private, so that unit-tests can inject a custom
+     * implementation.
+     */
+    static Logger log = LoggerFactory.getLogger(ValueConverter.class);
 
     private ValueConverter() {
     }
@@ -316,6 +327,9 @@ public final class ValueConverter {
      * <code>int</code>, and <code>short</code>, arrays of these primitive
      * types, {@link String}, and arrays of {@link String}.
      *
+     * @param pv_name
+     *  name of the process variable for which the value is converted. This is
+     *  only used for logging purposes.
      * @param object
      *  object to be converted.
      * @param charset
@@ -324,15 +338,20 @@ public final class ValueConverter {
      *  indicates whether a {@link String} or single element
      *  <code>String[]</code> array should be converted to a
      *  <code>DBR_CHAR</code> instead of a <code>DBR_STRING</code>.
+     * @param long_conversion_mode
+     *  behavior when converting a {@link Long} value that does not fit into an
+     *  {@link Integer}.
      * @return
      *  the converted value.
      * @throws IllegalArgumentException
      *  if <code>object</code> cannot be converted.
      */
     public static ChannelAccessSimpleOnlyValue<?> objectToChannelAccessSimpleOnlyValue(
+            String pv_name,
             Object object,
             Charset charset,
-            boolean convert_string_as_long_string) {
+            boolean convert_string_as_long_string,
+            JackiePreferences.LongConversionMode long_conversion_mode) {
         if (object instanceof VType vtype) {
             var converted_object = VTypeHelper.toObject(vtype);
             // VTypeHelper.toObject returns null if it does not know how to
@@ -367,6 +386,110 @@ public final class ValueConverter {
         if (object instanceof int[] value) {
             return ChannelAccessValueFactory.createLong(value);
         }
+        if (object instanceof Long value) {
+            if (value >= Integer.MIN_VALUE && value <= Integer.MAX_VALUE) {
+                return ChannelAccessValueFactory.createLong(
+                        new int[]{value.intValue()});
+            }
+            return switch (long_conversion_mode) {
+                case COERCE -> ChannelAccessValueFactory.createLong(
+                        new int[]{coerceToInt(value)});
+                case COERCE_AND_WARN -> {
+                    final var coerced_value = coerceToInt(value);
+                    log.warn(
+                            "Writing long {} as int {} for PV {}",
+                            value,
+                            coerced_value,
+                            pv_name);
+                    yield ChannelAccessValueFactory.createLong(
+                            new int[]{coerced_value});
+                }
+                case CONVERT -> ChannelAccessValueFactory.createDouble(
+                        new double[]{value.doubleValue()});
+                case CONVERT_AND_WARN -> {
+                    final var converted_value = value.doubleValue();
+                    log.warn(
+                            "Writing long {} as double {} for PV {}",
+                            value,
+                            converted_value,
+                            pv_name);
+                    yield ChannelAccessValueFactory.createDouble(
+                            new double[]{converted_value});
+                }
+                case FAIL -> throw new IllegalArgumentException(
+                        "Cannot write long value " + value
+                                + ". The value is outside the range of a "
+                                + "32-bit signed integer.");
+                case TRUNCATE -> ChannelAccessValueFactory.createLong(
+                        new int[]{value.intValue()});
+                case TRUNCATE_AND_WARN -> {
+                    final var truncated_value = value.intValue();
+                    log.warn(
+                            "Writing long {} as int {} for PV {}",
+                            value,
+                            truncated_value,
+                            pv_name);
+                    yield ChannelAccessValueFactory.createLong(
+                            new int[]{truncated_value});
+                }
+            };
+        }
+        if (object instanceof long[] value) {
+            boolean limit_exceeded = false;
+            for (long element : value) {
+                if (element < Integer.MIN_VALUE
+                        || element > Integer.MAX_VALUE) {
+                    limit_exceeded = true;
+                    break;
+                }
+            }
+            if (!limit_exceeded) {
+                return ChannelAccessValueFactory.createLong(
+                        longArrayToIntArray(value));
+            }
+            return switch (long_conversion_mode) {
+                case COERCE -> ChannelAccessValueFactory.createLong(
+                        coerceToInt(value));
+                case COERCE_AND_WARN -> {
+                    final var coerced_value = coerceToInt(value);
+                    log.warn(
+                            "Writing long[] {} as int[] {} for PV {}",
+                            Arrays.toString(value),
+                            Arrays.toString(coerced_value),
+                            pv_name);
+                    yield ChannelAccessValueFactory.createLong(
+                            coerced_value);
+                }
+                case CONVERT -> ChannelAccessValueFactory.createDouble(
+                        longArrayToDoubleArray(value));
+                case CONVERT_AND_WARN -> {
+                    final var converted_value = longArrayToDoubleArray(value);
+                    log.warn(
+                            "Writing long[] {} as double[] {} for PV {}",
+                            Arrays.toString(value),
+                            Arrays.toString(converted_value),
+                            pv_name);
+                    yield ChannelAccessValueFactory.createDouble(
+                            converted_value);
+                }
+                case FAIL -> throw new IllegalArgumentException(
+                        "Cannot write long[] value " + Arrays.toString(value)
+                                + ". The value is outside the range of a "
+                                + "32-bit signed integer.");
+                case TRUNCATE -> ChannelAccessValueFactory.createLong(
+                        longArrayToIntArray(value));
+                case TRUNCATE_AND_WARN -> {
+                    final var truncated_value = longArrayToIntArray(value);
+                    log.warn(
+                            "Writing long[] {} as int[] {} for PV {}",
+                            Arrays.toString(value),
+                            Arrays.toString(truncated_value),
+                            pv_name);
+                    yield ChannelAccessValueFactory.createLong(
+                            truncated_value);
+                }
+            };
+        }
         if (object instanceof Short value) {
             return ChannelAccessValueFactory.createShort(new short[] {value});
         }
@@ -389,7 +512,7 @@ public final class ValueConverter {
             // conversion if the array has a single element.
             if (value.length == 1 && convert_string_as_long_string) {
                 return objectToChannelAccessSimpleOnlyValue(
-                        value[0], charset, true);
+                        pv_name, value[0], charset, true, long_conversion_mode);
             }
             return ChannelAccessValueFactory.createString(
                     Arrays.asList(value), charset);
@@ -413,6 +536,21 @@ public final class ValueConverter {
                 return buffer.remaining();
             }
         };
+    }
+
+    private static int coerceToInt(long value) {
+        if (value > Integer.MAX_VALUE) {
+            return Integer.MAX_VALUE;
+        } else if (value < Integer.MIN_VALUE) {
+            return Integer.MIN_VALUE;
+        } else {
+            return (int) value;
+        }
+    }
+
+    private static int[] coerceToInt(long[] array) {
+        return Arrays.stream(array).mapToInt(
+                ValueConverter::coerceToInt).toArray();
     }
 
     private static Alarm convertAlarm(ChannelAccessTimeValue<?> time_value) {
@@ -537,6 +675,16 @@ public final class ValueConverter {
                 return buffer.remaining();
             }
         };
+    }
+
+    private static double[] longArrayToDoubleArray(long[] array) {
+        return Arrays.stream(array).mapToDouble(
+                element -> (double) element).toArray();
+    }
+
+    private static int[] longArrayToIntArray(long[] array) {
+        return Arrays.stream(array).mapToInt(
+                element -> (int) element).toArray();
     }
 
     private static ListShort shortBufferToListShort(ShortBuffer buffer) {

--- a/core/pv-jackie/src/main/resources/pv_jackie_preferences.properties
+++ b/core/pv-jackie/src/main/resources/pv_jackie_preferences.properties
@@ -119,6 +119,18 @@ honor_zero_precision=true
 # hostname is determined automatically.
 hostname=
 
+# How to process write operations specifying a long value that is greater than
+# 2147483647 or less than -2147483648. Possible settings are:
+#
+# - COERCE
+# - COERCE_AND_WARN
+# - CONVERT
+# - CONVERT_AND_WARN
+# - FAIL
+# - TRUNCATE
+# - TRUNCATE_AND_WARN
+long_conversion_mode=COERCE_AND_WARN
+
 # Mask that shall be used when registering monitors for DBR_TIME_* values.
 #
 # This can be a combination of DBE_ALARM, DBE_ARCHIVE, DBE_PROPERTY, and

--- a/core/pv-jackie/src/test/java/org/phoebus/pv/jackie/util/ValueConverterTest.java
+++ b/core/pv-jackie/src/test/java/org/phoebus/pv/jackie/util/ValueConverterTest.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2024 aquenos GmbH.
+ * Copyright (c) 2024-2025 aquenos GmbH.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -19,6 +19,7 @@ import com.aquenos.epics.jackie.common.value.ChannelAccessSimpleOnlyFloat;
 import com.aquenos.epics.jackie.common.value.ChannelAccessSimpleOnlyLong;
 import com.aquenos.epics.jackie.common.value.ChannelAccessSimpleOnlyShort;
 import com.aquenos.epics.jackie.common.value.ChannelAccessSimpleOnlyString;
+import com.aquenos.epics.jackie.common.value.ChannelAccessSimpleOnlyValue;
 import com.aquenos.epics.jackie.common.value.ChannelAccessTimeValue;
 import com.aquenos.epics.jackie.common.value.ChannelAccessValueFactory;
 import org.apache.commons.lang3.ArrayUtils;
@@ -42,6 +43,10 @@ import org.epics.vtype.VShortArray;
 import org.epics.vtype.VString;
 import org.epics.vtype.VStringArray;
 import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.mockito.verification.VerificationMode;
+import org.phoebus.pv.jackie.JackiePreferences;
+import org.slf4j.Logger;
 
 import java.nio.ByteBuffer;
 import java.nio.DoubleBuffer;
@@ -69,9 +74,7 @@ public class ValueConverterTest {
     @Test
     public void byteArrayToChannelAccessValue() {
         final byte[] value = new byte[] {2, 4};
-        final var ca_value = (ChannelAccessSimpleOnlyChar) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyChar) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(ByteBuffer.wrap(value), ca_value.getValue());
     }
 
@@ -81,9 +84,7 @@ public class ValueConverterTest {
     @Test
     public void byteToChannelAccessValue() {
         final byte value = 3;
-        final var ca_value = (ChannelAccessSimpleOnlyChar) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyChar) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(value, ca_value.getValue().get(0));
     }
 
@@ -555,9 +556,7 @@ public class ValueConverterTest {
     @Test
     public void doubleArrayToChannelAccessValue() {
         final double[] value = new double[] {2.0, 4.0};
-        final var ca_value = (ChannelAccessSimpleOnlyDouble) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyDouble) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(DoubleBuffer.wrap(value), ca_value.getValue());
     }
 
@@ -567,9 +566,7 @@ public class ValueConverterTest {
     @Test
     public void doubleToChannelAccessValue() {
         final double value = 3.0;
-        final var ca_value = (ChannelAccessSimpleOnlyDouble) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyDouble) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(value, ca_value.getValue().get(0));
     }
 
@@ -579,9 +576,7 @@ public class ValueConverterTest {
     @Test
     public void floatArrayToChannelAccessValue() {
         final float[] value = new float[] {2.0f, 4.0f};
-        final var ca_value = (ChannelAccessSimpleOnlyFloat) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyFloat) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(FloatBuffer.wrap(value), ca_value.getValue());
     }
 
@@ -591,9 +586,7 @@ public class ValueConverterTest {
     @Test
     public void floatToChannelAccessValue() {
         final float value = 3.0f;
-        final var ca_value = (ChannelAccessSimpleOnlyFloat) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyFloat) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(value, ca_value.getValue().get(0));
     }
 
@@ -603,9 +596,7 @@ public class ValueConverterTest {
     @Test
     public void intArrayToChannelAccessValue() {
         final int[] value = new int[] {2, 4};
-        final var ca_value = (ChannelAccessSimpleOnlyLong) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyLong) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(IntBuffer.wrap(value), ca_value.getValue());
     }
 
@@ -615,10 +606,73 @@ public class ValueConverterTest {
     @Test
     public void intToChannelAccessValue() {
         final int value = 3;
-        final var ca_value = (ChannelAccessSimpleOnlyLong) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyLong) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(value, ca_value.getValue().get(0));
+    }
+
+    /**
+     * Test conversion of a <code>long[]</code> to a CA value.
+     */
+    @Test
+    public void longArrayToChannelAccessValue() {
+        // Test an array that has a value that is too small.
+        var original = new long[] {15L, 23L, Integer.MIN_VALUE - 1L};
+        var coerced = new int[]{15, 23, Integer.MIN_VALUE};
+        var converted = new double[]{
+                15.0, 23.0, (double) (Integer.MIN_VALUE - 1L)};
+        var truncated = new int[]{15, 23, Integer.MAX_VALUE};
+        testLong(original, coerced, converted, truncated, false);
+        // Test an array that has a value that is too large.
+        original = new long[] {15L, 23L, Integer.MAX_VALUE + 1L};
+        coerced = new int[]{15, 23, Integer.MAX_VALUE};
+        converted = new double[]{15.0, 23.0, (double) (Integer.MAX_VALUE + 1L)};
+        truncated = new int[]{15, 23, Integer.MIN_VALUE};
+        testLong(original, coerced, converted, truncated, false);
+        // Test an array where all values are within the allowed range.
+        original = new long[] {15L, 23L, Integer.MIN_VALUE, Integer.MAX_VALUE};
+        coerced = new int[]{15, 23, Integer.MIN_VALUE, Integer.MAX_VALUE};
+        // converted is not used when all elements are within the limits.
+        converted = null;
+        truncated = coerced;
+        testLong(original, coerced, converted, truncated, true);
+    }
+
+    /**
+     * Test conversion of a {@link Long} to a CA value.
+     */
+    @Test
+    public void longToChannelAccessValue() {
+        // Test a value that is too small.
+        var original = Integer.MIN_VALUE - 1L;
+        var coerced = Integer.MIN_VALUE;
+        var converted = (double) (Integer.MIN_VALUE - 1L);
+        var truncated = Integer.MAX_VALUE;
+        testLong(original, coerced, converted, truncated, false);
+        // Test a value that is too large.
+        original = Integer.MAX_VALUE + 1L;
+        coerced = Integer.MAX_VALUE;
+        converted = (double) (Integer.MAX_VALUE + 1L);
+        truncated = Integer.MIN_VALUE;
+        testLong(original, coerced, converted, truncated, false);
+        // Test values in the allowed range. The converted value does not
+        // matter in this context, because it is never used.
+        original = 15L;
+        coerced = (int) original;
+        converted = Double.NaN;
+        truncated = (int) original;
+        testLong(original, coerced, converted, truncated, true);
+        original = 23L;
+        coerced = (int) original;
+        truncated = (int) original;
+        testLong(original, coerced, converted, truncated, true);
+        original = Integer.MIN_VALUE;
+        coerced = (int) original;
+        truncated = (int) original;
+        testLong(original, coerced, converted, truncated, true);
+        original = Integer.MAX_VALUE;
+        coerced = (int) original;
+        truncated = (int) original;
+        testLong(original, coerced, converted, truncated, true);
     }
 
     /**
@@ -627,9 +681,7 @@ public class ValueConverterTest {
     @Test
     public void shortArrayToChannelAccessValue() {
         final short[] value = new short[] {2, 4};
-        final var ca_value = (ChannelAccessSimpleOnlyShort) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyShort) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(ShortBuffer.wrap(value), ca_value.getValue());
     }
 
@@ -639,9 +691,7 @@ public class ValueConverterTest {
     @Test
     public void shortToChannelAccessValue() {
         final short value = 3;
-        final var ca_value = (ChannelAccessSimpleOnlyShort) ValueConverter
-                .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+        final var ca_value = (ChannelAccessSimpleOnlyShort) objectToChannelAccessSimpleOnlyValue(value);
         assertEquals(value, ca_value.getValue().get(0));
     }
 
@@ -653,19 +703,31 @@ public class ValueConverterTest {
         var value = new String[] {"abc", "123"};
         var ca_value = (ChannelAccessSimpleOnlyString) ValueConverter
                 .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+                        "",
+                        value,
+                        UTF_8,
+                        false,
+                        JackiePreferences.LongConversionMode.FAIL);
         assertEquals(Arrays.asList(value), ca_value.getValue());
         // For an array with multiple elements, it should not make a difference
         // if we enable the convert_string_as_long_string option.
         ca_value = (ChannelAccessSimpleOnlyString) ValueConverter
                 .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, true);
+                        "",
+                        value,
+                        UTF_8,
+                        true,
+                        JackiePreferences.LongConversionMode.FAIL);
         assertEquals(Arrays.asList(value), ca_value.getValue());
         // For a single-element array, we expect a different result.
         value = new String[] {"a single string"};
         final var ca_char_array = (ChannelAccessSimpleOnlyChar) ValueConverter
                 .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, true);
+                        "",
+                        value,
+                        UTF_8,
+                        true,
+                        JackiePreferences.LongConversionMode.FAIL);
         assertEquals(
                 value[0], UTF_8.decode(ca_char_array.getValue()).toString());
     }
@@ -678,13 +740,21 @@ public class ValueConverterTest {
         final String value = "some string";
         final var ca_value = (ChannelAccessSimpleOnlyString) ValueConverter
                 .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, false);
+                        "",
+                        value,
+                        UTF_8,
+                        false,
+                        JackiePreferences.LongConversionMode.FAIL);
         assertEquals(value, ca_value.getValue().get(0));
         // When setting convert_string_as_long_string to true, we expect a
         // ChannelAccessSimpleOnlyChar instead.
         final var ca_char_array = (ChannelAccessSimpleOnlyChar) ValueConverter
                 .objectToChannelAccessSimpleOnlyValue(
-                        value, UTF_8, true);
+                        "",
+                        value,
+                        UTF_8,
+                        true,
+                        JackiePreferences.LongConversionMode.FAIL);
         assertEquals(value, UTF_8.decode(ca_char_array.getValue()).toString());
     }
 
@@ -753,6 +823,12 @@ public class ValueConverterTest {
         assertEquals(controls_value.getLabels(),display.getChoices());
     }
 
+    private static void checkWarnCalled(
+            Logger logger_mock, VerificationMode verification_mode) {
+        Mockito.verify(logger_mock, verification_mode).warn(
+                Mockito.anyString(), (Object[]) Mockito.any());
+    }
+
     private static void checkTime(
             ChannelAccessTimeValue<?> time_value,
             TimeProvider time_provider) {
@@ -762,6 +838,162 @@ public class ValueConverterTest {
                         + ValueConverter.OFFSET_EPICS_TO_UNIX_EPOCH_SECONDS,
                 instant.getEpochSecond());
         assertEquals(time_value.getTimeNanoseconds(), instant.getNano());
+    }
+
+    private static ChannelAccessSimpleOnlyValue<?> objectToChannelAccessSimpleOnlyValue(
+            Object value) {
+        return ValueConverter.objectToChannelAccessSimpleOnlyValue(
+                "test_pv",
+                value,
+                UTF_8,
+                false,
+                JackiePreferences.LongConversionMode.FAIL);
+    }
+
+    private static ChannelAccessSimpleOnlyValue<?> objectToChannelAccessSimpleOnlyValue(
+            Object value,
+            JackiePreferences.LongConversionMode long_conversion_mode,
+            Logger logger) {
+        final var original_logger = ValueConverter.log;
+        try {
+            if (logger != null) {
+                ValueConverter.log = logger;
+            }
+            return ValueConverter.objectToChannelAccessSimpleOnlyValue(
+                    "test_pv",
+                    value,
+                    UTF_8,
+                    false,
+                    long_conversion_mode);
+        } finally {
+            if (logger != null) {
+                ValueConverter.log = original_logger;
+            }
+        }
+    }
+
+    private static void testLong(
+            long original,
+            int coerced,
+            double converted,
+            int truncated,
+            boolean in_limits) {
+        testLong(
+                original,
+                new int[]{coerced},
+                new double[]{converted},
+                new int[]{truncated},
+                in_limits);
+    }
+
+    private static void testLong(
+            Object original,
+            int[] coerced,
+            double[] converted,
+            int[] truncated,
+            boolean in_limits) {
+        // Test the COERCE mode.
+        var logger_mock = Mockito.mock(Logger.class);
+        var ca_value = objectToChannelAccessSimpleOnlyValue(
+                original,
+                JackiePreferences.LongConversionMode.COERCE,
+                logger_mock);
+        assertEquals(
+                IntBuffer.wrap(coerced),
+                ((ChannelAccessSimpleOnlyLong) ca_value).getValue());
+        checkWarnCalled(logger_mock, Mockito.never());
+        // Test the COERCE_AND_WARN mode.
+        logger_mock = Mockito.mock(Logger.class);
+        ca_value = objectToChannelAccessSimpleOnlyValue(
+                original,
+                JackiePreferences.LongConversionMode.COERCE_AND_WARN,
+                logger_mock);
+        assertEquals(
+                IntBuffer.wrap(coerced),
+                ((ChannelAccessSimpleOnlyLong) ca_value).getValue());
+        if (in_limits) {
+            checkWarnCalled(logger_mock, Mockito.never());
+        } else {
+            checkWarnCalled(logger_mock, Mockito.times(1));
+        }
+        // Test the CONVERT mode.
+        logger_mock = Mockito.mock(Logger.class);
+        ca_value = objectToChannelAccessSimpleOnlyValue(
+                original,
+                JackiePreferences.LongConversionMode.CONVERT, logger_mock);
+        if (in_limits) {
+            assertEquals(
+                    IntBuffer.wrap(truncated),
+                    ((ChannelAccessSimpleOnlyLong) ca_value).getValue());
+            checkWarnCalled(logger_mock, Mockito.never());
+        } else {
+            assertEquals(
+                    DoubleBuffer.wrap(converted),
+                    ((ChannelAccessSimpleOnlyDouble) ca_value).getValue());
+            checkWarnCalled(logger_mock, Mockito.never());
+        }
+        // Test the CONVERT_AND_WARN mode.
+        logger_mock = Mockito.mock(Logger.class);
+        ca_value = objectToChannelAccessSimpleOnlyValue(
+                original,
+                JackiePreferences.LongConversionMode.CONVERT_AND_WARN,
+                logger_mock);
+        if (in_limits) {
+            assertEquals(
+                    IntBuffer.wrap(truncated),
+                    ((ChannelAccessSimpleOnlyLong) ca_value).getValue());
+            checkWarnCalled(logger_mock, Mockito.never());
+        } else {
+            assertEquals(
+                    DoubleBuffer.wrap(converted),
+                    ((ChannelAccessSimpleOnlyDouble) ca_value).getValue());
+            checkWarnCalled(logger_mock, Mockito.times(1));
+        }
+        // Test the FAIL mode.
+        logger_mock = Mockito.mock(Logger.class);
+        if (in_limits) {
+            ca_value = objectToChannelAccessSimpleOnlyValue(
+                    original,
+                    JackiePreferences.LongConversionMode.FAIL,
+                    logger_mock);
+            assertEquals(
+                    IntBuffer.wrap(truncated),
+                    ((ChannelAccessSimpleOnlyLong) ca_value).getValue());
+            checkWarnCalled(logger_mock, Mockito.never());
+        } else {
+            final var final_logger_mock = logger_mock;
+            assertThrows(
+                    IllegalArgumentException.class,
+                    () -> objectToChannelAccessSimpleOnlyValue(
+                            original,
+                            JackiePreferences.LongConversionMode.FAIL,
+                            final_logger_mock));
+            checkWarnCalled(logger_mock, Mockito.never());
+        }
+        // Test the TRUNCATE mode.
+        logger_mock = Mockito.mock(Logger.class);
+        ca_value = objectToChannelAccessSimpleOnlyValue(
+                original,
+                JackiePreferences.LongConversionMode.TRUNCATE,
+                logger_mock);
+        assertEquals(
+                IntBuffer.wrap(truncated),
+                ((ChannelAccessSimpleOnlyLong) ca_value).getValue());
+        checkWarnCalled(logger_mock, Mockito.never());
+        // Test the TRUNCATE_AND_WARN mode.
+        logger_mock = Mockito.mock(Logger.class);
+        ca_value = objectToChannelAccessSimpleOnlyValue(
+                original,
+                JackiePreferences.LongConversionMode.TRUNCATE_AND_WARN,
+                logger_mock);
+        assertEquals(
+                IntBuffer.wrap(truncated),
+                ((ChannelAccessSimpleOnlyLong) ca_value).getValue());
+        if (in_limits) {
+            checkWarnCalled(logger_mock, Mockito.never());
+        } else {
+            checkWarnCalled(logger_mock, Mockito.times(1));
+        }
     }
 
 }

